db/batch/kv_cache.go

```diff
@@ -6,96 +6,170 @@
-		deleted: make(map[hash.Hash160]struct{}),
+	return &kvCache{
+		cache:   make(map[string]map[string][]byte),
+		deleted: make(map[string]map[string]struct{}),
 	}
-	return &c
 }
 
 // Read retrieves a record
-func (c *kvCache) Read(k hash.Hash160) ([]byte, error) {
-	if v, ok := c.cache[k]; ok {
+func (c *kvCache) Read(key *KvCacheKey) ([]byte, error) {
+	if v := c.cacheGet(key); v != nil {
 		return v, nil
 	}
-	if _, ok := c.deleted[k]; ok {
+	if c.deletedGet(key) {
 		return nil, ErrAlreadyDeleted
 	}
 	return nil, ErrNotExist
 }
 
 // Write puts a record into cache
-func (c *kvCache) Write(k hash.Hash160, v []byte) {
-	c.cache[k] = v
-	delete(c.deleted, k)
+func (c *kvCache) Write(key *KvCacheKey, v []byte) {
+	c.cachePut(key, v)
+	c.deletedDelete(key)
 }
 
 // WriteIfNotExist puts a record into cache only if it doesn't exist, otherwise return ErrAlreadyExist
-func (c *kvCache) WriteIfNotExist(k hash.Hash160, v []byte) error {
-	if _, ok := c.cache[k]; ok {
+func (c *kvCache) WriteIfNotExist(key *KvCacheKey, v []byte) error {
+	if v := c.cacheGet(key); v != nil {
 		return ErrAlreadyExist
 	}
-	c.cache[k] = v
-	delete(c.deleted, k)
+	c.Write(key, v)
 	return nil
 }
 
 // Evict deletes a record from cache
-func (c *kvCache) Evict(k hash.Hash160) {
-	delete(c.cache, k)
-	c.deleted[k] = struct{}{}
+func (c *kvCache) Evict(key *KvCacheKey) {
+	c.cacheDelete(key)
+	c.deletedPut(key)
 }
 
 // Clear clear the cache
 func (c *kvCache) Clear() {
-	c.cache = nil
-	c.deleted = nil
-	c.cache = make(map[hash.Hash160][]byte)
-	c.deleted = make(map[hash.Hash160]struct{})
+	c.cacheClear()
+	c.deletedClear()
+	c.cache = make(map[string]map[string][]byte)
+	c.deleted = make(map[string]map[string]struct{})
 }
 
 // Clone clones the cache
 func (c *kvCache) Clone() KVStoreCache {
 	clone := kvCache{
-		cache:   make(map[hash.Hash160][]byte),
-		deleted: make(map[hash.Hash160]struct{}),
+		cache:   make(map[string]map[string][]byte),
+		deleted: make(map[string]map[string]struct{}),
 	}
 	// clone entries in map
-	for k, v := range c.cache {
-		clone.cache[k] = v
+	for k1, v1 := range c.cache {
+		clone.cache[k1] = make(map[string][]byte)
+		for k2, v2 := range v1 {
+			clone.cache[k1][k2] = v2
+		}
 	}
-	for k, v := range c.deleted {
-		clone.deleted[k] = v
+	for k1, v1 := range c.deleted {
+		clone.deleted[k1] = make(map[string]struct{})
+		for k2, v2 := range v1 {
+			clone.deleted[k1][k2] = v2
+		}
 	}
 	return &clone
 }
+
+func (c *kvCache) cachePut(key *KvCacheKey, v []byte) {
+	if _, ok := c.cache[key.key1]; !ok {
+		c.cache[key.key1] = make(map[string][]byte)
+	}
+	c.cache[key.key1][key.key2] = v
+}
+
+func (c *kvCache) cacheGet(key *KvCacheKey) []byte {
+	if v1, ok := c.cache[key.key1]; ok {
+		if v2, ok := v1[key.key2]; ok {
+			return v2
+		}
+	}
+	return nil
+}
+
+func (c *kvCache) cacheDelete(key *KvCacheKey) {
+	if v1, ok := c.cache[key.key1]; ok {
+		delete(v1, key.key2)
+		if len(v1) == 0 {
+			delete(c.cache, key.key1)
+		}
```

Comment:
actually don't need to delete? if `v1` becomes an empty map, `cacheGet` continues to work correctly, while at the same time it avoids make() of a new map inside `cachePut`

Reply:
It should. Otherwise, it becomes zoombie in the map